### ISSUE TITLE
Internal error: Failed to run on the given Interpreter: tensorflow/lite/kernels/reduce.cc:534 

### ISSUE THREAD

**Issue by Sudhan97** [tensorflow/tensorflow#46010]
**Descriptions**
We successfully converted the Matterpot Mask RCNN object detection model (using hdf5 weight file) into tflite format using tensorflow (version 2.3.0). We got the prediction from the converted tflite model using python API of tf-nightly () due to Flex delegate issue raised when using tensorflow version 2.3.0.

**Converted tflite model link** : https://drive.google.com/file/d/1kJjQnuf5FYdn0DcEAIkeMoWFXZFuM2eb/view?usp=sharing

When we tried to get the prediction using our converted tflite model using the android application (https://github.com/tensorflow/examples/tree/master/lite/examples/object_detection/android) but the following error is raised at **runForMultipleInputsOutputs()** function.

**Error raised in android :**
```
2020-12-28 11:30:01.263 6838-6857/org.tensorflow.lite.examples.detection E/AndroidRuntime: FATAL EXCEPTION: inference
    Process: org.tensorflow.lite.examples.detection, PID: 6838
    java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: tensorflow/lite/kernels/reduce.cc:534 reference_ops::ReduceGeneric<T>( GetTensorData<T>(op_context->input), op_context->input->dims->data, op_context->input->dims->size, GetTensorData<T>(op_context->output), op_context->output->dims->data, op_context->output->dims->size, GetTensorData<int>(op_context->axis), num_axis, op_context->params->keep_dims, GetTensorData<int>(temp_index), GetTensorData<int>(resolved_axis), init_value, reducer) was not true.
    No
        at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:158)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:343)
        at org.tensorflow.lite.examples.detection.tflite.TFLiteObjectDetectionAPIModel.recognizeImage(TFLiteObjectDetectionAPIModel.java:334)
        at org.tensorflow.lite.examples.detection.DetectorActivity$2.run(DetectorActivity.java:182)
        at android.os.Handler.handleCallback(Handler.java:761)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:156)
        at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Used python code used for converting the model into tflite format**
```
model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR, config=config)
keras_model = `model.keras_model`
converter = tf.lite.TFLiteConverter.from_keras_model(keras_model)
converter.allow_custom_ops = True
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
    tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
]
converter.optimizations = [ tf.lite.Optimize.DEFAULT ]
tflite_model = converter.convert()
```

**Java code used in android studio to the get the prediction**
We tested the code for single image and set input tensors accordly (images, img_metas and anchors).

```
// File:  org.tensorflow.lite.examples.detection.tflite.TFLiteObjectDetectionAPIModel

Interpreter.Options options = new Interpreter.Options();

// Model file is loaded from assets.
Interpreter Tflite = new Interpreter(modelFile, options);

float[][][][] images = new float[1][1024][1024][3];
float[][] img_metas = new float[1][14];
float[][][] anchors = new float[1][261888][4];
// above arrays are populated. 

Object[] inputArray = {images, img_metas, anchors};
tensor0 = new float[1][1][1];
tensor1 = new float[1][1000][2][4];
tensor2 = new float[1][1000][2];
tensor3 = new float[1][100][6];
tensor4 = new float[1][100][28][28][2];
tensor5 = new float[1][1][4];
tensor6 = new float[1][1][2];
Map<Integer, Object> outputMap = new HashMap<>();
    outputMap.put(0, tensor0);
    outputMap.put(1, tensor1);
    outputMap.put(2, tensor2);
    outputMap.put(3, tensor3);
    outputMap.put(4, tensor4);
    outputMap.put(5, tensor5);
    outputMap.put(6, tensor6);
tfLite.runForMultipleInputsOutputs(inputArray, outputMap);
```

**Comment by thaink** [comment]
Can you try our latest version 2.4.0 for both conversion and inference?
I tried to run your model but it seemed to run successfully.

**Comment by Sudhan97** [comment]
@thaink Error is not raised in Python API but in android application (Java code). Following are the tensorflow lite dependencies used in the android application with above mentioned model.
```
    // Build off of nightly TensorFlow Lite
    implementation('org.tensorflow:tensorflow-lite:0.0.0-nightly') { changing = true }
    implementation 'org.tensorflow:tensorflow-lite-select-tf-ops:0.0.0-nightly'
    implementation 'org.tensorflow:tensorflow-lite-support:0.0.0-nightly'
    implementation ('org.tensorflow:tensorflow-lite-metadata:0.0.0-nightly') { changing = true }
```

**Comment by abattery** [comment]
@Sudhan97 could you share the error message with the tf-nightly version in the android application?

**Comment by Sudhan97** [comment]
> @Sudhan97 could you share the error message on the tf-nightly version in the android application?

@abattery Above I have included the **error raised in android**, **java code used android application** and **python code we used to convert into tflite format**.
Anyway I'll include the error message again here.

```
2020-12-28 11:30:01.263 6838-6857/org.tensorflow.lite.examples.detection E/AndroidRuntime: FATAL EXCEPTION: inference
    Process: org.tensorflow.lite.examples.detection, PID: 6838
    java.lang.IllegalArgumentException: Internal error: Failed to run on the given Interpreter: tensorflow/lite/kernels/reduce.cc:534 reference_ops::ReduceGeneric<T>( GetTensorData<T>(op_context->input), op_context->input->dims->data, op_context->input->dims->size, GetTensorData<T>(op_context->output), op_context->output->dims->data, op_context->output->dims->size, GetTensorData<int>(op_context->axis), num_axis, op_context->params->keep_dims, GetTensorData<int>(temp_index), GetTensorData<int>(resolved_axis), init_value, reducer) was not true.
    No
        at org.tensorflow.lite.NativeInterpreterWrapper.run(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:158)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:343)
        at org.tensorflow.lite.examples.detection.tflite.TFLiteObjectDetectionAPIModel.recognizeImage(TFLiteObjectDetectionAPIModel.java:334)
        at org.tensorflow.lite.examples.detection.DetectorActivity$2.run(DetectorActivity.java:182)
        at android.os.Handler.handleCallback(Handler.java:761)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:156)
        at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by abattery** [comment]
I suspect that the attached error is not coming from the tf-nightly version. The line number is different with the tf-nightly version.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/reduce.cc#L558

Could you confirm that your TFLite dependency is based on TF 2.5.0 in the android application?

FYI, in the recent version, we have a fix for dealing with empty reduction inputs.

**Comment by Sudhan97** [comment]
@abattery thank you, I used the latest version 2.4 and I was able get the prediction. But when using the latest stable version(2.4.0) for converting the model to TFlite format, the converted model take around a hour for single image prediction output. Model converted using version 2.3.1 maximum took around 2 mins for prediction ouput.

**Comment by Sudhan97** [comment]
@abattery @thaink Thank you, we are able to get the prediction using latest tensorflow version 2.4.0. 
I'm closing this issue

**Comment by MeghaGhosh** [comment]
Hi @Sudhan97 I have followed the same steps you have outlined here. However I could do with some help understanding the outputs of the model. The details are here https://github.com/tensorflow/tensorflow/issues/46951

Which of the 7 output tensors provide the bounding box values? For me the values keep getting repeated for each detection and I don't seem to see any values that resemble the bounding box.

**Comment by MeghaGhosh** [comment]
Hey @Sudhan97 how did you get the prediction in tf-nightly? I've tried many times and it always results in segmentation fault.

**Comment by Sudhan97** [comment]
@MeghaGhosh Hi, Sorry for this late reply. Output tensor 3 is detection and 4 is the masks. When I used tf-nightly(dev 2.5), I didn't get any segmentation errors. Try using tensorflow 2.4.0 version

**Comment by MeghaGhosh** [comment]
Thanks @Sudhan97 yes that finally worked for us. We were making a mistake with passing the image meta. I could not get it to run on any version of the python API though, even used a docker image. Anyway that is fine.

So the thing is, depending on how we resize/pad the image we always get different predictions. Using the tensorflow lite support library for image processing got us the closest to the keras h5 predictions, but the result is still a bit weird. Do you mind sharing which method you used for image resize in android and whether you got results consistent with the h5 file?
